rfcs/hse/0004-kvdb-kvs-introspection-apis.md

```diff
@@ -0,0 +1,156 @@
+ */
+const char *
+hse_kvs_name_get(const struct hse_kvs *kvs);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvdb_param_get(const struct hse_kvdb *kvdb, const char *param, char *buf, size_t buf_len);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvs_param_get(const struct hse_kvs *kvs, const char *param, char *buf, size_t buf_len);
+
+enum hse_mclass {
+  HSE_MCLASS_CAPACITY,
+  HSE_MCLASS_STAGING,
+};
+
+struct hse_mclass_info {
+  uint64_t total_bytes;     /**< total space in the file-system containing this media class */
+  uint64_t available_bytes; /**< available space in the file-system containing this media class */
+  uint64_t allocated_bytes; /**< allocated storage space for a media class */
+  uint64_t used_bytes;      /**< used storage space for a media class */
+  char     path[PATH_MAX];  /**< path to media class */
+};
+
+/* Populates the above storage info struct with given information.
+ *
+ * Con: Since we only track data owned by each media class, we drop the storage
+ * statistics about the kvdb.meta file, but I don't see that as being super
+ * important.
+ */
+hse_err_t
+hse_kvdb_mclass_info_get(const struct hse_kvdb *kvdb, enum hse_mclass mclass, struct hse_mclass_info *info);
+```
+
+#### REST APIs
+
+One issue we have currently that is worth solving in this RFC is that our REST
+API assumes there is only one KVDB. We seem to have settled on the idea we want
+to support more than one KVDB in the future. Now is the perfect time to retrofit
+the REST API.
+
+One issue with name-less KVDBs is that file paths look like parts of a REST
+path, which is not good for easy usage. Typically in a well-planned REST route
+tree, you would have something like `/kvdb/<specific kvdb>`, where `/kvdb` is
+the parent route of all KVDBs. In our current, name-less KVDB situation this
+would look like `/kvdb/path/to/kvdb` where `path/to/kvdb` is the KVDB home. This
+makes it impossible to parse REST paths correctly. What we can do instead is the
+following:
+
+- Add a notion of a REST alias for a KVDB
+  - The alias would default to `0` for the first KVDB opened, `1` for the next
+    KVDB opened and so on. In the event KVDB 0 is closed and re-opened, KVDB 0
+    would become `2`.
+
+The following REST APIs will be added:
+
+- `/kvdb`
+  - `GET` - Return a JSON body of all currently opened KVDBs
+- `/kvdb/<alias>/home`
+  - `GET` - Return a JSON body of the KVDB home
+- `/kvdb/<alias>/mclass/<mclass: "capacity" | "staging">/info`
+  - `GET` - Return a JSON body of the mclass info struct
+- `/kvdb/<alias>/params`
+  - `GET` - Return a JSON body of all the KVDB parameters
+- `/kvdb/<alias>/params/<param>`
+  - `GET` - Return a JSON body of just the parameter that was asked for
+  - `PUT` - Change the value of a parameter which is WRITABLE
+  - `DELETE` - Set the value of a parameter which is WRITABLE to its default
```

Comment:
How about a way to find the default value of a parameter?  One idea would be to add structure to output of `GET`, for example: `{ "current": 7, "default": 3 }`.  And, if we go there, you could add `"origin": <string>` to indicate where the current value came from (default, paramv, config file, REST).  That would be useful to me as a developer when looking at a system to understand what's going on.

Reply:
I am not entirely sure I see the value in showing the defaults, It seems like an implementation detail.

I don't find the origin of the parameter as necessary info. The origin is currently noted within debug logging however if one was curious.

Reply:
When someone says, "hey, something's not right on my system, can you take a look?", the first thing I want to do is know what parameters have been changed from their default settings.   You might see a prefix or pivot length that was too high, or a space amp setting that was a bit aggressive.  We have many settings that will completely change system behavior if tweaked too far from defaults.  Being able to see what is different from defaults is valuable IMO.

If you don't like mucking up the output JSON body, it could be an option to the query.  For example: `/kvdb/<alias>/params?default=true` or something similar.

The origin is less valuable.

Reply:
The default query parameter sounds like a good compromise to me. Will add it to RFC.